### PR TITLE
feat: Add Wholesale services request period length validation

### PR DESCRIPTION
**Pull request by MWO1024**
# Description

## Pull-request quality

- [ ] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [ ] Tests are written and executed locally
- [ ] Subsystem tests have been tested (by manually deploying to `dev_002` or `sandbox_002`)
- [ ] Documentation has been updated
- [ ] The integration [event catalog](https://energinet.atlassian.net/wiki/spaces/D3/pages/555581556/Event+catalog) has been updated
- [ ] C4 diagrams have been updated and Team Outlaws has been informed about relevant changes

### CODE REVIEW

File: source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs [hunk 1]
```diff
@@ -125,8 +150,21 @@ public async Task Validate_WhenPeriodStartIsOlderThanAllowed_ReturnsExpectedVali
...
         // Arrange
+        _now = new LocalDateTime(2024, 6, 1, 0, 0, 0)
+            .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
+            .ToInstant();
+
+        var start = new LocalDateTime(2021, 4, 1, 0, 0, 0)
+            .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
+            .ToInstant();
+
+        var end = new LocalDateTime(_now.ToDateTimeOffset().Year, 5, 1, 0, 0, 0)
```
Comment (ebbeknudsen) [inline source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs:161]:
```suggestion
        var end = new LocalDateTime(2024, 5, 1, 0, 0, 0)
```

File: source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs [hunk 2]
```diff
@@ -192,32 +240,52 @@ public async Task Validate_WhenPeriodStartIsExactly3Years2MonthsMinus1HourOldDue
...
         // Act
         var errors = await _sut.ValidateAsync(message);
 
         // Assert
         using var assertionScope = new AssertionScope();
-        errors.Should().ContainSingle();
-        var error = errors.First();
-        error.ErrorCode.Should().Be(_startDateMustBeLessThanOrEqualTo3YearsAnd2Months.ErrorCode);
-        error.Message.Should().Be(_startDateMustBeLessThanOrEqualTo3YearsAnd2Months.Message);
+        errors.OrderBy(e => e.Message).Should().SatisfyRespectively(
```
Comment (ebbeknudsen) [inline source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs:266]: Why `OrderBy` and `SatisfyRespectively`? Don't we only expect 1 validation error, like the other tests?

File: source/dotnet/wholesale-api/Edi/Validation/WholesaleServicesRequest/Rules/PeriodValidationRule.cs [hunk 3]
```diff
@@ -13,12 +13,12 @@
...
 
 using NodaTime;
-using NodaTime.Extensions;
 using NodaTime.Text;
 
 namespace Energinet.DataHub.Wholesale.Edi.Validation.WholesaleServicesRequest.Rules;
 
-public sealed class PeriodValidationRule : IValidationRule<DataHub.Edi.Requests.WholesaleServicesRequest>
+public sealed class PeriodValidationRule(DateTimeZone dateTimeZone, IClock clock)
+    : IValidationRule<DataHub.Edi.Requests.WholesaleServicesRequest>
```
Comment (ebbeknudsen) [inline source/dotnet/wholesale-api/Edi/Validation/WholesaleServicesRequest/Rules/PeriodValidationRule.cs:21]: Shouldn't this reuse some of the logic from the `PeriodValidationRule` for `AggregatedTimeSeriesRequest`? As far as i remember one of the reasons we chose FluentValidation was to be able to reuse validation rules

File: source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs [hunk 4]
```diff
@@ -144,12 +182,17 @@ public async Task Validate_WhenPeriodStartIsExactly3Years2MonthsAnd1HourOldDueTo
             .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
             .ToInstant();
 
+        var periodEndDate = new LocalDateTime(2021, 11, 1, 0, 0, 0)
+            .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
```
Comment (lasrinnil) [inline source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs:186]: Consider moving the `Europe/Copenhagen` to a constant

File: source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs [hunk 5]
```diff
@@ -170,12 +213,17 @@ public async Task Validate_WhenPeriodStartIsExactly3Years2MonthsMinus1HourOldDue
             .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
             .ToInstant();
 
+        var periodEndDate = new LocalDateTime(2021, 4, 1, 0, 0, 0)
+            .InZoneStrictly(DateTimeZoneProviders.Tzdb.GetZoneOrNull("Europe/Copenhagen")!)
```
Comment (lasrinnil) [inline source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs:217]: Could we use a `SystemClock` and configure its timezone to `Europe/Copenhagen` that way we wouldn't  have to specify it in every test

**Review by lasrinnil — DISMISSED** [review]

File: source/dotnet/wholesale-api/Edi/Validation/WholesaleServicesRequest/Rules/PeriodValidationRule.cs (hunk 3 above)
Comment (MadsDue) [reply source/dotnet/wholesale-api/Edi/Validation/WholesaleServicesRequest/Rules/PeriodValidationRule.cs:22]: Yes it sould. 

I have introduced a simple PeriodValidationHelper, which contains the shared logic.

File: source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs (hunk 5 above)
Comment (MadsDue) [reply source/dotnet/wholesale-api/Edi.UnitTests/Validators/WholesaleServicesRequest/PeriodValidationRuleTests.cs:214]: It looks like IClock is returning an Instant which is not localized.

File: source/dotnet/wholesale-api/Edi/Validation/Helpers/PeriodValidationHelper.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+public class PeriodValidationHelper(DateTimeZone dateTimeZone, IClock clock)
+{
+    public bool IsMidnight(Instant instant, out ZonedDateTime zonedDateTime)
+    {
+        zonedDateTime = new ZonedDateTime(instant, dateTimeZone);
+
+        return zonedDateTime.TimeOfDay == LocalTime.Midnight;
+    }
+
+    public bool IsStartDateOlderTheAllowed(Instant start, int maxYears, int maxMonths)
```
Comment (ebbeknudsen) [inline source/dotnet/wholesale-api/Edi/Validation/Helpers/PeriodValidationHelper.cs:28]:
```suggestion
    public bool IsStartDateOlderThanAllowed(Instant start, int maxYears, int maxMonths)
```

File: source/dotnet/wholesale-api/Edi/Validation/Helpers/PeriodValidationHelper.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        var zonedStartDateTime = new ZonedDateTime(start, dateTimeZone);
+        var zonedCurrentDateTime = new ZonedDateTime(clock.GetCurrentInstant(), dateTimeZone);
+        var latestStartDate = zonedCurrentDateTime.LocalDateTime.PlusYears(-maxYears).PlusMonths(maxMonths);
+
+        return zonedStartDateTime.LocalDateTime < latestStartDate;
+    }
+
+    public bool IntervalMustBeLessThenAllowedPeriodSize(Instant start, Instant end, int maxAllowedPeriodSizeInMonths)
```
Comment (ebbeknudsen) [inline source/dotnet/wholesale-api/Edi/Validation/Helpers/PeriodValidationHelper.cs:37]:
```suggestion
    public bool IntervalMustBeLessThanAllowedPeriodSize(Instant start, Instant end, int maxAllowedPeriodSizeInMonths)
```

**Review by ebbeknudsen — APPROVED** [review]

**Review by ebbeknudsen — APPROVED** [review]

**Review by ebbeknudsen — APPROVED** [review]

**Pull request merged**